File: lua/SledgehammerLua/media/lua/client/Sledgehammer/Modules/Factions/Objects/FactionMember.lua
-- This file is part of Sledgehammer.
--
--    Sledgehammer is free software: you can redistribute it and/or modify
--    it under the terms of the GNU Lesser General Public License as published by
--    the Free Software Foundation, either version 3 of the License, or
--    (at your option) any later version.
--
--    Sledgehammer is distributed in the hope that it will be useful,
--    but WITHOUT ANY WARRANTY; without even the implied warranty of
--    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
--    GNU Lesser General Public License for more details.
--
--    You should have received a copy of the GNU Lesser General Public License
--    along with Sledgehammer. If not, see <http://www.gnu.org/licenses/>.
--
--	Sledgehammer is free to use and modify, ONLY for non-official third-party servers 
--    not affiliated with TheIndieStone, or it's immediate affiliates, or contractors. 

require "Sledgehammer"

----------------------------------------------------------------
-- FactionMember.lua
-- 
-- @module Factions
-- @author Jab
-- @license LGPL
--
-- @number id 		The internal ID of the Player.
-- @string name 	The name of the Player.
-- @number rank 	The rank flag of the Player. 
----------------------------------------------------------------
FactionMember = class(function(o, id, name, rank)
	
	-- Set the variables.
	o.id      = id;
	o.name    = name;
	o.rank    = rank;
	o.faction = nil;
end);

----------------------------------------------------------------
-- @return 	Returns the internal ID of the Player.
----------------------------------------------------------------
function FactionMember:getID()
	return self.id;
end

----------------------------------------------------------------
-- Sets the internal ID of the Player.
--
-- @private
-- @string ID 	The internal ID of the Player.
----------------------------------------------------------------
function FactionMember:setID(ID)
	self.ID = ID;
end

----------------------------------------------------------------
-- @return 	Returns the name of the Player.
----------------------------------------------------------------
function FactionMember:getName()
	return self.name;
end

----------------------------------------------------------------
-- @return Returns the Rank flag of the Player.
----------------------------------------------------------------
function FactionMember:getRank()
	return self.rank;
end

----------------------------------------------------------------
-- @return Returns the Faction the Player is a member of.
----------------------------------------------------------------
function FactionMember:getFaction()
	return self.faction;
end

----------------------------------------------------------------
-- Sets the Faction the Player is a member of.
--
-- @private
-- @Faction faction 	The Faction the Player is a member of.
----------------------------------------------------------------
function FactionMember:setFaction(faction)
	self.faction = faction;
end

File: src/main/java/sledgehammer/lua/faction/FactionMember.java
/*
 * This file is part of Sledgehammer.
 *
 *    Sledgehammer is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU Lesser General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    Sledgehammer is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public License
 *    along with Sledgehammer. If not, see <http://www.gnu.org/licenses/>.
 *
 *    Sledgehammer is free to use and modify, ONLY for non-official third-party servers
 *    not affiliated with TheIndieStone, or it's immediate affiliates, or contractors.
 */

package sledgehammer.lua.faction;

import java.util.UUID;

import se.krka.kahlua.vm.KahluaTable;
import sledgehammer.SledgeHammer;
import sledgehammer.database.module.faction.MongoFactionMember;
import sledgehammer.lua.MongoLuaObject;
import sledgehammer.lua.core.Color;
import sledgehammer.lua.core.Player;
import sledgehammer.module.faction.ModuleFactions;

/**
 * MongoLuaObject to handle faction-member data and operations for the Factions Module.
 *
 * @author Jab
 */
public class FactionMember extends MongoLuaObject<MongoFactionMember> {

  /** The Faction the FactionMember is in. */
  private Faction faction;

  /**
   * Main constructor.
   *
   * @param mongoDocument The MongoDocument container.
   */
  public FactionMember(MongoFactionMember mongoDocument) {
    super(mongoDocument, "FactionMember");
    setMongoDocument(mongoDocument);
  }

  @Override
  public void onLoad(KahluaTable table) {
    // TODO: implement.
  }

  @Override
  public void onExport() {
    // TODO: implement.
  }

  /**
   * Sets the new Faction.
   *
   * @param faction The Faction to set.
   * @param save The flag to save the Document.
   */
  public void setFaction(Faction faction, boolean save) {
    Faction factionOld = getFaction();
    if (factionOld != null) {
      factionOld.removeMember(this);
    }
    if (faction != null) {
      boolean success = faction.addMember(this);
      if (success) {
        this.faction = faction;
        getMongoDocument().setFactionId(faction.getUniqueId(), save);
      }
    } else {
      this.faction = null;
    }
    Player player = SledgeHammer.instance.getPlayer(getPlayerId());
    setTag(player);
  }

  public void setTag(Player player) {
    if (player != null) {
      if (faction != null) {
        player.setNickname("[" + faction.getFactionTag() + "] " + player.getUsername());
        player.setColor(Color.getColor(faction.getFactionRawColor()));
      } else {
        player.setNickname(null);
        player.setColor(Color.WHITE);
      }
    }
  }

  /** Removes the FactionMember from the Faction. */
  public void leaveFaction() {
    ModuleFactions moduleFactions = SledgeHammer.instance.getModule(ModuleFactions.class);
    moduleFactions.removeFactionMember(this);
  }

  /** @return Returns the Unique ID of the Player as the FactionMember. */
  public UUID getPlayerId() {
    return getMongoDocument().getPlayerId();
  }

  /** @return Returns the Unique ID of the Faction the FactionMember is in. */
  public UUID getFactionId() {
    return getMongoDocument().getFactionId();
  }

  /** @return Returns the Faction the FactionMember is in. */
  public Faction getFaction() {
    return this.faction;
  }
}
